be/src/util/failpoint/fail_point.cpp

```diff
@@ -0,0 +1,179 @@
+    // If there is a performance problem here, we can switch back to fiu's own interfaces one by one.
+    fiu_enable_external(name.c_str(), 1, nullptr, 0, check_fail_point);
+    return Status::OK();
+}
+
+FailPoint* FailPointRegistry::get(const std::string& name) {
+    auto iter = _fps.find(name);
+    if (iter == _fps.end()) {
+        return nullptr;
+    }
+    return iter->second;
+}
+
+bool init_failpoint_from_conf(const std::string& conf_file) {
+    if (!std::filesystem::exists(conf_file.c_str())) {
+        return true;
+    }
+    try {
+        LOG(INFO) << "load failpoint from config file: " << conf_file;
+        // load file
+        simdjson::ondemand::parser parser;
+        auto json = simdjson::padded_string::load(conf_file.c_str());
+        simdjson::ondemand::document doc = parser.iterate(json);
+        auto object = doc.get_object();
+        for (auto field : object) {
+            auto fp_name = field.unescaped_key();
+            if (fp_name.error() != simdjson::SUCCESS) {
+                LOG(WARNING) << "cannot parse json key from config file";
+                return false;
+            }
+            auto fp = FailPointRegistry::GetInstance()->get(std::string(fp_name.value()));
+            if (fp == nullptr) {
+                LOG(WARNING) << "cannot find failpoint with name " << fp_name.value();
+                return false;
+            }
+            auto value = field.value();
+            auto mode = value["mode"].get_string();
+            if (mode.error() != simdjson::SUCCESS) {
+                return false;
+            }
+            PFailPointTriggerMode trigger_mode;
+            if (mode.value() == "enable") {
+                trigger_mode.set_mode(FailPointTriggerModeType::ENABLE);
+            } else if (mode.value() == "enable_n_times") {
+                auto n_times = value["n_times"].get_int64();
+                if (n_times.error() != simdjson::SUCCESS) {
+                    return false;
+                }
+                trigger_mode.set_mode(FailPointTriggerModeType::ENABLE_N_TIMES);
+                trigger_mode.set_n_times(n_times.value());
+            } else if (mode.value() == "probability_enable") {
+                auto probability = value["probability"].get_double();
+                if (probability.error() != simdjson::SUCCESS) {
+                    return false;
+                }
+                trigger_mode.set_mode(FailPointTriggerModeType::PROBABILITY_ENABLE);
+                trigger_mode.set_probability(probability.value());
+            }
+            fp->setMode(trigger_mode);
+        }
+        return true;
+    } catch (...) {
+    }
+    return false;
+}
+
+FailPointRegistry::FailPointRegistry() {
+#ifdef FIU_ENABLE
+    fiu_init(0);
+#endif
+}
+
+FailPointRegisterer::FailPointRegisterer(FailPoint* fp) {
+    FailPointRegistry::GetInstance()->add(fp);
+}
+
+DEFINE_FAIL_POINT(random_error);
```

Comment:
> At present, only one fp? can we define more failpoints? during testing, some are on, some are off. suggest to provide a statement to show all the failpoints.
> 
> ```
> show failpoints;
> show failpoints like "%random%";
> ```

1. at present, we have two failpoints named `mem_alloc_error` and `random_error`. `mem_alloc_error` is used to control whether memory allocation should return an error.`random_error` is a temporary failpoint, used to control whether some interfaces should return an error status.
2. I use `random_error` just for convenience, because only one failpoint setting needs to be changed during test. The goal at this stage is to make the random fault injection test run, and then I will split it into multiple failpoints.
3. we can of course define multiple failpoints and are encouraged to do so.
4. I have add a command to show failpoints, plz see the demo in description.